File: 2022/Day03/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>

using namespace std;

int getDuplicateItem(string line)
{
	const int count = line.length() / 2;

	for (int i = 0; i < count; i++)
	{
		char toFind = line[i];
		if (line.find(toFind, count) != string::npos)
			return toFind;
	}
	return -1;
	
}

bool isCapital(const char c)
{
	return (c >= 65 && c <= 90);
}

int getScore(char c)
{
	if (isCapital(c))
		return (c - 'A' + 27);
	else
		return (c - 'a' + 1);
}

// one string split in two, find common character
void ex01(vector<string> &backpacks)
{
	int const size = backpacks.size();
	int prioritySum = 0;

	for(int i = 0; i < size; i++)
	{	
		string &line = backpacks[i];
		char item = getDuplicateItem(line);
		prioritySum += getScore(item);
	}

	cout << "ex01: " << prioritySum << endl;
}

//  common char across three strings
void ex02(vector<string> &backpacks)
{
	int const size = backpacks.size();
	int prioritySum = 0;

	for(int j = 0; j < size; j += 3)
	{
		string &b = backpacks[j + 1];
		string &c = backpacks[j + 2];
		for (int i = 0; i < backpacks[j].size(); i++)
		{
			char item = backpacks[j][i];
			if (b.find(item, 0) != string::npos && c.find(item, 0) != string::npos) 
			{
				prioritySum += getScore(item);
				break ;
			}
		}
	}
	cout << "ex02: " << prioritySum << endl;
}

int main(void)
{
	fstream inputFile;
	stringstream buffer;
	
	inputFile.open("input.txt", ios::in);
	
	buffer << inputFile.rdbuf();
	vector<string> backpacks;
	for(string line; getline(buffer, line);)
		backpacks.push_back(line);

	ex01(backpacks);
	ex02(backpacks);

	return 0;
}

File: 2022/Day03/main.py
# import array as arr

def ex01(lines):
    prioritySum = 0
    for line in lines:
        size = int(len(line) / 2)
        string1, string2 = line[0:size], line[size:]
        commonchar = set(string1).intersection(string2)
        for item in commonchar:
            if item.islower():
                prioritySum += ord(item) - ord('a') + 1
            else:
                prioritySum += ord(item) - ord('A') + 27

    print("ex01: ", prioritySum)


def ex02(lines):
    team = []
    prioritySum = 0
    for i, line in enumerate(lines):
        if i % 3 == 0:
            team.clear()
        team.append(line)
        if i > 0 and i % 3 == 2:
            common = set(team[0]) & set(team[1]) & set(team[2])
            for item in common:
                if item.islower():
                    prioritySum += ord(item) - ord('a') + 1
                else:
                    prioritySum += ord(item) - ord('A') + 27

    print("ex02: ", prioritySum)


def __main__():
    input = open("input.txt", 'r')
    lines = input.read().splitlines()
    ex01(lines)
    ex02(lines)


__main__()
